### ISSUE TITLE
`NamedTuple`-specific methods not part of auto-completions

### ISSUE THREAD

**Issue by sharkdp** [astral-sh/ty#1161]
Start with:

```py
from typing import NamedTuple


class Person(NamedTuple):
    name: str
    age: int | None = None

person = Person("Alice", 25)
```

and then try to auto-complete on `person.<CURSOR>`. This should include methods from [`NamedTupleFallback`](https://github.com/python/typeshed/blob/de819980041bf9f86ba120e1f070456ce8800d5a/stdlib/_typeshed/_type_checker_internals.pyi#L55), e.g. `_asdict`, but does not.

I *think* this should be easy to fix by doing something similar to what we do for `TypedDict`s [here](https://github.com/astral-sh/ruff/blob/9cb37db5101771eb08f2e90f5b556b33d73a1308/crates/ty_python_semantic/src/types/ide_support.rs#L102-L112), only with `NamedTuple` and `NamedTupleFallback`.

**Comment by swordreforge** [comment]
**Problem Analysis:**

The issue is that when a class is defined inheriting from `typing.NamedTuple`, the created instance (`person`) lacks the methods specific to named tuples (like `_asdict`, `_replace`, `_fields`) in its auto-completion list provided by the language server (e.g., Pylance using Pyright).

**Root Cause:**
The type checker's resolver likely interprets `class Person(NamedTuple)` as a simple class with fields `name` and `age` but fails to properly synthesize its type as a subclass of `tuple`. Consequently, it doesn't inherit the type information containing the special methods provided by `NamedTuple`.

**Proposed Solution:**

You are absolutely correct. The fix involves implementing a special case for `NamedTuple` similar to the existing one for `TypedDict`.

Here is a step-by-step solution:

1.  **Identify the Code Location:** This fix needs to be implemented within the static type checker's (e.g., Pyright's) source code, specifically in the part that handles class definition resolution and inheritance.

2.  **Detect NamedTuple Inheritance:** The parser needs a special branch to detect the pattern when a class inherits from `typing.NamedTuple`.

3.  **Synthesize the Correct Type:** Upon identifying a `NamedTuple` class, the type checker should not treat it as a regular class. Instead, it must **synthesize** a new type. This synthesized type should be a combination (e.g., an **Intersection** type) of two components:
    *   An **Anonymous Tuple Type:** This represents the user-defined fields (`name: str`, `age: int | None`), enabling attribute access (e.g., `person.name`).
    *   The **NamedTuple Fallback Type:** This is a predefined type representing the `tuple` base class and includes all the `NamedTuple`-specific methods (`_asdict`, `_replace`, `_fields`, etc.) and regular `tuple` methods (`index`, `count`).

    The instance type of `Person` should be set to this combined, synthesized type.

4.  **Implement Completion Provider:** In the language server's auto-completion module, when it encounters `person.`, it must query this synthesized type. It will then gather attributes from both components:
    *   `name` and `age` from the anonymous tuple type.
    *   `_asdict`, `_replace`, `index`, `count`, etc., from the fallback type.
    This combined list is then presented to the user for auto-completion.

**Pseudocode Example:**

The implementation in the type checker's internals would look something like this:

```python
# This is inside the type checker's code, not user code.

def resolve_class_definition(class_node):
    base_types = get_base_types(class_node)

    # Check if the class inherits from typing.NamedTuple
    if any(is_namedtuple_base(base) for base in base_types):
        # 1. Create an anonymous type for the user-defined fields
        field_type = create_anonymous_record_type(class_node)
        
        # 2. Get the standard fallback type for NamedTuple (contains _asdict, _replace, etc.)
        namedtuple_fallback_type = get_namedtuple_fallback_type()
        
        # 3. Combine the types (e.g., using an Intersection type)
        synthesized_type = combine_types(field_type, namedtuple_fallback_type)
        
        return synthesized_type

    # ... Other handling for regular classes, TypedDict, etc. ...
    else:
        return resolve_regular_class(class_node)
```
In summary, the solution requires the type checker to perform **specialized handling** for `NamedTuple`, constructing a synthetic type that is a hybrid of a defined object and a tuple, ensuring the language server can offer completions for both fields and methods.
